File: majorProject/src/main/java/com/aocc/framework/implementation/AndroidFastRenderView.java
package com.aocc.framework.implementation;

import android.graphics.Canvas;
import android.graphics.Color;
import android.view.Choreographer;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import com.aocc.framework.GameConstants;
import com.aocc.framework.Viewport;

/**
 * Renders each frame directly to the device surface at native resolution.
 * Game logic and layout use virtual world coordinates; the viewport transform
 * scales them to the letterboxed on-screen area.
 */
public class AndroidFastRenderView extends SurfaceView implements SurfaceHolder.Callback {
    private final AndroidGame game;
    private final AndroidGraphics graphics;
    private final SurfaceHolder holder;
    private final Choreographer choreographer = Choreographer.getInstance();
    private final Choreographer.FrameCallback frameCallback = this::onFrame;

    private volatile boolean running;
    private long lastFrameTimeNanos;

    public AndroidFastRenderView(AndroidGame game, AndroidGraphics graphics) {
        super(game);
        this.game = game;
        this.graphics = graphics;
        this.holder = getHolder();
        holder.addCallback(this);
    }

    public void resume() {
        running = true;
        lastFrameTimeNanos = System.nanoTime();
        choreographer.postFrameCallback(frameCallback);
    }

    public void pause() {
        running = false;
        choreographer.removeFrameCallback(frameCallback);
    }

    private void onFrame(long frameTimeNanos) {
        if (!running) {
            return;
        }

        if (!holder.getSurface().isValid()) {
            choreographer.postFrameCallback(frameCallback);
            return;
        }

        float deltaSeconds = (frameTimeNanos - lastFrameTimeNanos) / 1_000_000_000f;
        lastFrameTimeNanos = frameTimeNanos;
        if (deltaSeconds > GameConstants.MAX_DELTA_SECONDS) {
            deltaSeconds = GameConstants.MAX_DELTA_SECONDS;
        }
        if (deltaSeconds < 0f) {
            deltaSeconds = 0f;
        }

        Canvas canvas = holder.lockCanvas();
        if (canvas != null) {
            try {
                Viewport viewport = game.getViewport();
                viewport.update(canvas.getWidth(), canvas.getHeight());
                game.updateInputViewport(viewport);

                game.getCurrentScreen().update(deltaSeconds);

                canvas.drawColor(Color.BLACK);
                graphics.beginFrame(canvas, viewport);
                try {
                    game.getCurrentScreen().paint(deltaSeconds);
                } finally {
                    graphics.endFrame();
                }
            } catch (RuntimeException e) {
                running = false;
                throw e;
            } finally {
                holder.unlockCanvasAndPost(canvas);
            }
        }

        choreographer.postFrameCallback(frameCallback);
    }

    @Override
    public void surfaceCreated(SurfaceHolder surfaceHolder) {
        if (running) {
            lastFrameTimeNanos = System.nanoTime();
            choreographer.postFrameCallback(frameCallback);
        }
    }

    @Override
    public void surfaceChanged(SurfaceHolder surfaceHolder, int format, int width, int height) {
        Viewport viewport = game.getViewport();
        viewport.update(width, height);
        game.updateInputViewport(viewport);
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder surfaceHolder) {
        choreographer.removeFrameCallback(frameCallback);
    }
}

File: majorProject/src/main/kotlin/com/aocc/framework/implementation/AndroidFastRenderView.kt
package com.aocc.framework.implementation

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.view.Choreographer
import android.view.SurfaceHolder
import android.view.SurfaceView
import com.aocc.framework.GameConstants
import com.aocc.framework.Viewport

/**
 * Game loop: update → paint into a letterbox-sized off-screen buffer → single
 * blit to the native surface. Avoids per-primitive surface scaling cost.
 */
class AndroidFastRenderView(
    private val game: AndroidGame,
    private val graphics: AndroidGraphics
) : SurfaceView(game), SurfaceHolder.Callback {

    private val holder: SurfaceHolder = getHolder()
    private val choreographer: Choreographer = Choreographer.getInstance()
    private val frameCallback: Choreographer.FrameCallback = Choreographer.FrameCallback { frameTimeNanos ->
        onFrame(frameTimeNanos)
    }
    private val blitPaint = Paint(Paint.FILTER_BITMAP_FLAG)

    @Volatile
    private var running: Boolean = false
    private var lastFrameTimeNanos: Long = 0

    init {
        holder.addCallback(this)
    }

    fun resume() {
        running = true
        lastFrameTimeNanos = System.nanoTime()
        choreographer.postFrameCallback(frameCallback)
    }

    fun pause() {
        running = false
        choreographer.removeFrameCallback(frameCallback)
    }

    private fun onFrame(frameTimeNanos: Long) {
        if (!running) {
            return
        }

        if (!holder.surface.isValid) {
            choreographer.postFrameCallback(frameCallback)
            return
        }

        var deltaSeconds = (frameTimeNanos - lastFrameTimeNanos) / 1_000_000_000f
        lastFrameTimeNanos = frameTimeNanos
        if (deltaSeconds > GameConstants.MAX_DELTA_SECONDS) {
            deltaSeconds = GameConstants.MAX_DELTA_SECONDS
        }
        if (deltaSeconds < 0f) {
            deltaSeconds = 0f
        }

        game.currentScreen.update(deltaSeconds)

        val canvas = holder.lockCanvas()
        if (canvas != null) {
            try {
                val viewport = game.viewport
                viewport.update(canvas.width, canvas.height)
                game.updateInputViewport(viewport)

                if (graphics.beginFrame(viewport)) {
                    try {
                        game.currentScreen.paint(deltaSeconds)
                    } finally {
                        graphics.endFrame()
                    }
                }

                canvas.drawColor(Color.BLACK)
                val frameBuffer = graphics.frameBuffer
                if (frameBuffer != null) {
                    val dest = viewport.getLetterboxDestRect()
                    canvas.drawBitmap(frameBuffer, null, dest, blitPaint)
                }
            } catch (e: RuntimeException) {
                running = false
                throw e
            } finally {
                holder.unlockCanvasAndPost(canvas)
            }
        }

        choreographer.postFrameCallback(frameCallback)
    }

    override fun surfaceCreated(surfaceHolder: SurfaceHolder) {
        if (running) {
            lastFrameTimeNanos = System.nanoTime()
            choreographer.postFrameCallback(frameCallback)
        }
    }

    override fun surfaceChanged(surfaceHolder: SurfaceHolder, format: Int, width: Int, height: Int) {
        val viewport = game.viewport
        viewport.update(width, height)
        game.updateInputViewport(viewport)
        graphics.ensureFramebuffer(viewport)
    }

    override fun surfaceDestroyed(surfaceHolder: SurfaceHolder) {
        choreographer.removeFrameCallback(frameCallback)
    }
}
